UnicodeDecodeError raised by log forwarder

Occurred 7 times within 25 minutes on 2023/09/29
Link to [CloudWatch logs](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#logsV2:logs-insights$3FqueryDetail$3D~(end~'2023-09-30T06*3a59*3a58.198Z~start~'2023-09-30T05*3a59*3a04.157Z~timeType~'ABSOLUTE~tz~'Local~editorString~'fields*20*40timestamp*2c*20*40message*0a*7c*20filter*20*40message*20like*20*2fERROR*2f*0a*7c*20sort*20*40timestamp*20asc*0a*7c*20limit*20200~queryId~'b84ff2e8-ba0f-4344-9766-51f55980897b~source~(~'*2faws*2flambda*2fazul-indexer-prod-forward_s3_logs)))

```
[ERROR] UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 420: ordinal not in range(128)
Traceback (most recent call last):
  File "/var/task/azul/chalice.py", line 166, in patched_event_source_handler
    return old_handler(self_, event, context)
  File "/var/task/chalice/app.py", line 1752, in __call__
    return self.handler(event_obj)
  File "/var/task/chalice/app.py", line 1708, in __call__
    return self.handler(request, self.next_handler)
  File "/var/task/azul/chalice.py", line 191, in _lambda_context_middleware
    return get_response(event)
  File "/var/task/app.py", line 293, in forward_s3_logs
    app.log_controller.forward_s3_access_logs(event)
  File "/var/task/azul/indexer/log_forwarding_controller.py", line 47, in forward_s3_access_logs
    self._forward_logs(event, self.s3)
  File "/var/task/azul/indexer/log_forwarding_controller.py", line 38, in _forward_logs
    for message in service.read_logs(event.bucket, event.key):
  File "/var/task/azul/indexer/log_forwarding_service.py", line 38, in read_logs
    for message in self._parse_log_lines(body):
  File "/var/task/azul/indexer/log_forwarding_service.py", line 130, in _parse_log_lines
    for message in super()._parse_log_lines(file_body):
  File "/var/task/azul/indexer/log_forwarding_service.py", line 54, in _parse_log_lines
    for row in csv.reader(file_body, delimiter=' ', quotechar='"'):
  File "/var/task/azul/indexer/log_forwarding_service.py", line 127, in _read_log
    yield line.decode('ascii')
```

This happened again over this past weekend in `anvilprod`, 

**CloudWatch Logs Insights**    
region: us-east-1    
log-group-names: /aws/lambda/azul-indexer-anvilprod-forward_s3_logs    
start-time: 2024-12-14T18:20:00.000Z    
end-time: 2024-12-14T22:32:30.000Z    
query-string:
  ```
  fields @timestamp, @message
| filter @message like 'DecodeError'
| sort @timestamp desc
| limit 10
  ```
---
| @timestamp | @message |
| --- | --- |
| 2024-12-14 22:29:07.666 | [ERROR] UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 315: ordinal not in range(128) Traceback (most recent call last):   File "/var/task/azul/chalice.py", line 181, in patched_event_source_handler     return old_handler(self_, event, context)   File "/var/task/chalice/app.py", line 1756, in __call__     return self.handler(event_obj)   File "/var/task/app.py", line 255, in forward_s3_logs     app.log_controller.forward_s3_access_logs(event)   File "/var/task/azul/indexer/log_forwarding_controller.py", line 47, in forward_s3_access_logs     self._forward_logs(event, self.s3)   File "/var/task/azul/indexer/log_forwarding_controller.py", line 38, in _forward_logs     for message in service.read_logs(event.bucket, event.key):   File "/var/task/azul/indexer/log_forwarding_service.py", line 38, in read_logs     for message in self._parse_log_lines(body):   File "/var/task/azul/indexer/log_forwarding_service.py", line 130, in _parse_log_lines     for message in super()._parse_log_lines(file_body):   File "/var/task/azul/indexer/log_forwarding_service.py", line 54, in _parse_log_lines     for row in csv.reader(file_body, delimiter=' ', quotechar='"'):   File "/var/task/azul/indexer/log_forwarding_service.py", line 127, in _read_log     yield line.decode('ascii') |
| 2024-12-14 22:27:02.481 | [ERROR] UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 315: ordinal not in range(128) Traceback (most recent call last):   File "/var/task/azul/chalice.py", line 181, in patched_event_source_handler     return old_handler(self_, event, context)   File "/var/task/chalice/app.py", line 1756, in __call__     return self.handler(event_obj)   File "/var/task/app.py", line 255, in forward_s3_logs     app.log_controller.forward_s3_access_logs(event)   File "/var/task/azul/indexer/log_forwarding_controller.py", line 47, in forward_s3_access_logs     self._forward_logs(event, self.s3)   File "/var/task/azul/indexer/log_forwarding_controller.py", line 38, in _forward_logs     for message in service.read_logs(event.bucket, event.key):   File "/var/task/azul/indexer/log_forwarding_service.py", line 38, in read_logs     for message in self._parse_log_lines(body):   File "/var/task/azul/indexer/log_forwarding_service.py", line 130, in _parse_log_lines     for message in super()._parse_log_lines(file_body):   File "/var/task/azul/indexer/log_forwarding_service.py", line 54, in _parse_log_lines     for row in csv.reader(file_body, delimiter=' ', quotechar='"'):   File "/var/task/azul/indexer/log_forwarding_service.py", line 127, in _read_log     yield line.decode('ascii') |
| 2024-12-14 22:26:33.125 | [ERROR] UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 316: ordinal not in range(128) Traceback (most recent call last):   File "/var/task/azul/chalice.py", line 181, in patched_event_source_handler     return old_handler(self_, event, context)   File "/var/task/chalice/app.py", line 1756, in __call__     return self.handler(event_obj)   File "/var/task/app.py", line 255, in forward_s3_logs     app.log_controller.forward_s3_access_logs(event)   File "/var/task/azul/indexer/log_forwarding_controller.py", line 47, in forward_s3_access_logs     self._forward_logs(event, self.s3)   File "/var/task/azul/indexer/log_forwarding_controller.py", line 38, in _forward_logs     for message in service.read_logs(event.bucket, event.key):   File "/var/task/azul/indexer/log_forwarding_service.py", line 38, in read_logs     for message in self._parse_log_lines(body):   File "/var/task/azul/indexer/log_forwarding_service.py", line 130, in _parse_log_lines     for message in super()._parse_log_lines(file_body):   File "/var/task/azul/indexer/log_forwarding_service.py", line 54, in _parse_log_lines     for row in csv.reader(file_body, delimiter=' ', quotechar='"'):   File "/var/task/azul/indexer/log_forwarding_service.py", line 127, in _read_log     yield line.decode('ascii') |
| 2024-12-14 22:25:55.304 | [ERROR] UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 315: ordinal not in range(128) Traceback (most recent call last):   File "/var/task/azul/chalice.py", line 181, in patched_event_source_handler     return old_handler(self_, event, context)   File "/var/task/chalice/app.py", line 1756, in __call__     return self.handler(event_obj)   File "/var/task/app.py", line 255, in forward_s3_logs     app.log_controller.forward_s3_access_logs(event)   File "/var/task/azul/indexer/log_forwarding_controller.py", line 47, in forward_s3_access_logs     self._forward_logs(event, self.s3)   File "/var/task/azul/indexer/log_forwarding_controller.py", line 38, in _forward_logs     for message in service.read_logs(event.bucket, event.key):   File "/var/task/azul/indexer/log_forwarding_service.py", line 38, in read_logs     for message in self._parse_log_lines(body):   File "/var/task/azul/indexer/log_forwarding_service.py", line 130, in _parse_log_lines     for message in super()._parse_log_lines(file_body):   File "/var/task/azul/indexer/log_forwarding_service.py", line 54, in _parse_log_lines     for row in csv.reader(file_body, delimiter=' ', quotechar='"'):   File "/var/task/azul/indexer/log_forwarding_service.py", line 127, in _read_log     yield line.decode('ascii') |
| 2024-12-14 22:24:35.924 | [ERROR] UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 316: ordinal not in range(128) Traceback (most recent call last):   File "/var/task/azul/chalice.py", line 181, in patched_event_source_handler     return old_handler(self_, event, context)   File "/var/task/chalice/app.py", line 1756, in __call__     return self.handler(event_obj)   File "/var/task/app.py", line 255, in forward_s3_logs     app.log_controller.forward_s3_access_logs(event)   File "/var/task/azul/indexer/log_forwarding_controller.py", line 47, in forward_s3_access_logs     self._forward_logs(event, self.s3)   File "/var/task/azul/indexer/log_forwarding_controller.py", line 38, in _forward_logs     for message in service.read_logs(event.bucket, event.key):   File "/var/task/azul/indexer/log_forwarding_service.py", line 38, in read_logs     for message in self._parse_log_lines(body):   File "/var/task/azul/indexer/log_forwarding_service.py", line 130, in _parse_log_lines     for message in super()._parse_log_lines(file_body):   File "/var/task/azul/indexer/log_forwarding_service.py", line 54, in _parse_log_lines     for row in csv.reader(file_body, delimiter=' ', quotechar='"'):   File "/var/task/azul/indexer/log_forwarding_service.py", line 127, in _read_log     yield line.decode('ascii') |
| 2024-12-14 22:23:36.745 | [ERROR] UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 316: ordinal not in range(128) Traceback (most recent call last):   File "/var/task/azul/chalice.py", line 181, in patched_event_source_handler     return old_handler(self_, event, context)   File "/var/task/chalice/app.py", line 1756, in __call__     return self.handler(event_obj)   File "/var/task/app.py", line 255, in forward_s3_logs     app.log_controller.forward_s3_access_logs(event)   File "/var/task/azul/indexer/log_forwarding_controller.py", line 47, in forward_s3_access_logs     self._forward_logs(event, self.s3)   File "/var/task/azul/indexer/log_forwarding_controller.py", line 38, in _forward_logs     for message in service.read_logs(event.bucket, event.key):   File "/var/task/azul/indexer/log_forwarding_service.py", line 38, in read_logs     for message in self._parse_log_lines(body):   File "/var/task/azul/indexer/log_forwarding_service.py", line 130, in _parse_log_lines     for message in super()._parse_log_lines(file_body):   File "/var/task/azul/indexer/log_forwarding_service.py", line 54, in _parse_log_lines     for row in csv.reader(file_body, delimiter=' ', quotechar='"'):   File "/var/task/azul/indexer/log_forwarding_service.py", line 127, in _read_log     yield line.decode('ascii') |
---

The following are the execution logs associated with a single request ID, which includes various retry attempts.
**CloudWatch Logs Insights**    
region: us-east-1    
log-group-names: /aws/lambda/azul-indexer-anvilprod-forward_s3_logs    
start-time: 2024-12-14T22:25:28.774Z    
end-time: 2024-12-14T22:29:30.173Z    
query-string:
  ```
  fields @timestamp, @message
| filter @logStream like '2024/12/14/[$LATEST]6c40bea745084a5798872564cd60272c'
| filter @message like /7ee248cf|ERROR|LAMBDA_WARNING/
| sort @timestamp asc
| limit 29
  ```
---
| @timestamp | @message |
| --- | --- |
| 2024-12-14 22:25:55.276 | START RequestId: 7ee248cf-0988-48e1-971e-999432d214a6 Version: $LATEST |
| 2024-12-14 22:25:55.278 | [INFO] 2024-12-14T22:25:55.278Z 7ee248cf-0988-48e1-971e-999432d214a6 azul.boto3 s3.GetObject: Making GET request to https://s3.amazonaws.com/edu-ucsc-gi-platform-anvil-prod-logs.us-east-1/s3/access/anvilprod/browser/2024-12-14-22-25-54-E363A95D4EFEAC8A |
| 2024-12-14 22:25:55.284 | [INFO] 2024-12-14T22:25:55.278Z 7ee248cf-0988-48e1-971e-999432d214a6 azul.boto3 s3.GetObject: … without request body |
| 2024-12-14 22:25:55.303 | [INFO] 2024-12-14T22:25:55.303Z 7ee248cf-0988-48e1-971e-999432d214a6 azul.boto3 s3.GetObject: Got 200 response |
| 2024-12-14 22:25:55.303 | [INFO] 2024-12-14T22:25:55.303Z 7ee248cf-0988-48e1-971e-999432d214a6 azul.boto3 s3.GetObject: … with nonprintable body (<class 'botocore.response.StreamingBody'>) |
| 2024-12-14 22:25:55.304 | LAMBDA_WARNING: Unhandled exception. The most likely cause is an issue in the function code. However, in rare cases, a Lambda runtime update can cause unexpected function behavior. For functions using managed runtimes, runtime updates can be triggered by a function change, or can be applied automatically. To determine if the runtime has been updated, check the runtime version in the INIT_START log entry. If this error correlates with a change in the runtime version, you may be able to mitigate this error by temporarily rolling back to the previous runtime version. For more information, see https://docs.aws.amazon.com/lambda/latest/dg/runtimes-update.html |
| 2024-12-14 22:25:55.304 | [ERROR] UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 315: ordinal not in range(128) Traceback (most recent call last):   File "/var/task/azul/chalice.py", line 181, in patched_event_source_handler     return old_handler(self_, event, context)   File "/var/task/chalice/app.py", line 1756, in __call__     return self.handler(event_obj)   File "/var/task/app.py", line 255, in forward_s3_logs     app.log_controller.forward_s3_access_logs(event)   File "/var/task/azul/indexer/log_forwarding_controller.py", line 47, in forward_s3_access_logs     self._forward_logs(event, self.s3)   File "/var/task/azul/indexer/log_forwarding_controller.py", line 38, in _forward_logs     for message in service.read_logs(event.bucket, event.key):   File "/var/task/azul/indexer/log_forwarding_service.py", line 38, in read_logs     for message in self._parse_log_lines(body):   File "/var/task/azul/indexer/log_forwarding_service.py", line 130, in _parse_log_lines     for message in super()._parse_log_lines(file_body):   File "/var/task/azul/indexer/log_forwarding_service.py", line 54, in _parse_log_lines     for row in csv.reader(file_body, delimiter=' ', quotechar='"'):   File "/var/task/azul/indexer/log_forwarding_service.py", line 127, in _read_log     yield line.decode('ascii') |
| 2024-12-14 22:25:55.305 | END RequestId: 7ee248cf-0988-48e1-971e-999432d214a6 |
| 2024-12-14 22:25:55.305 | REPORT RequestId: 7ee248cf-0988-48e1-971e-999432d214a6 Duration: 28.93 ms Billed Duration: 29 ms Memory Size: 128 MB Max Memory Used: 122 MB |
| 2024-12-14 22:26:33.104 | LAMBDA_WARNING: Unhandled exception. The most likely cause is an issue in the function code. However, in rare cases, a Lambda runtime update can cause unexpected function behavior. For functions using managed runtimes, runtime updates can be triggered by a function change, or can be applied automatically. To determine if the runtime has been updated, check the runtime version in the INIT_START log entry. If this error correlates with a change in the runtime version, you may be able to mitigate this error by temporarily rolling back to the previous runtime version. For more information, see https://docs.aws.amazon.com/lambda/latest/dg/runtimes-update.html |
| 2024-12-14 22:26:33.125 | [ERROR] UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 316: ordinal not in range(128) Traceback (most recent call last):   File "/var/task/azul/chalice.py", line 181, in patched_event_source_handler     return old_handler(self_, event, context)   File "/var/task/chalice/app.py", line 1756, in __call__     return self.handler(event_obj)   File "/var/task/app.py", line 255, in forward_s3_logs     app.log_controller.forward_s3_access_logs(event)   File "/var/task/azul/indexer/log_forwarding_controller.py", line 47, in forward_s3_access_logs     self._forward_logs(event, self.s3)   File "/var/task/azul/indexer/log_forwarding_controller.py", line 38, in _forward_logs     for message in service.read_logs(event.bucket, event.key):   File "/var/task/azul/indexer/log_forwarding_service.py", line 38, in read_logs     for message in self._parse_log_lines(body):   File "/var/task/azul/indexer/log_forwarding_service.py", line 130, in _parse_log_lines     for message in super()._parse_log_lines(file_body):   File "/var/task/azul/indexer/log_forwarding_service.py", line 54, in _parse_log_lines     for row in csv.reader(file_body, delimiter=' ', quotechar='"'):   File "/var/task/azul/indexer/log_forwarding_service.py", line 127, in _read_log     yield line.decode('ascii') |
| 2024-12-14 22:27:02.460 | START RequestId: 7ee248cf-0988-48e1-971e-999432d214a6 Version: $LATEST |
| 2024-12-14 22:27:02.461 | [INFO] 2024-12-14T22:27:02.461Z 7ee248cf-0988-48e1-971e-999432d214a6 azul.boto3 s3.GetObject: Making GET request to https://s3.amazonaws.com/edu-ucsc-gi-platform-anvil-prod-logs.us-east-1/s3/access/anvilprod/browser/2024-12-14-22-25-54-E363A95D4EFEAC8A |
| 2024-12-14 22:27:02.461 | [INFO] 2024-12-14T22:27:02.461Z 7ee248cf-0988-48e1-971e-999432d214a6 azul.boto3 s3.GetObject: … without request body |
| 2024-12-14 22:27:02.481 | [INFO] 2024-12-14T22:27:02.481Z 7ee248cf-0988-48e1-971e-999432d214a6 azul.boto3 s3.GetObject: Got 200 response |
| 2024-12-14 22:27:02.481 | [INFO] 2024-12-14T22:27:02.481Z 7ee248cf-0988-48e1-971e-999432d214a6 azul.boto3 s3.GetObject: … with nonprintable body (<class 'botocore.response.StreamingBody'>) |
| 2024-12-14 22:27:02.481 | LAMBDA_WARNING: Unhandled exception. The most likely cause is an issue in the function code. However, in rare cases, a Lambda runtime update can cause unexpected function behavior. For functions using managed runtimes, runtime updates can be triggered by a function change, or can be applied automatically. To determine if the runtime has been updated, check the runtime version in the INIT_START log entry. If this error correlates with a change in the runtime version, you may be able to mitigate this error by temporarily rolling back to the previous runtime version. For more information, see https://docs.aws.amazon.com/lambda/latest/dg/runtimes-update.html |
| 2024-12-14 22:27:02.481 | [ERROR] UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 315: ordinal not in range(128) Traceback (most recent call last):   File "/var/task/azul/chalice.py", line 181, in patched_event_source_handler     return old_handler(self_, event, context)   File "/var/task/chalice/app.py", line 1756, in __call__     return self.handler(event_obj)   File "/var/task/app.py", line 255, in forward_s3_logs     app.log_controller.forward_s3_access_logs(event)   File "/var/task/azul/indexer/log_forwarding_controller.py", line 47, in forward_s3_access_logs     self._forward_logs(event, self.s3)   File "/var/task/azul/indexer/log_forwarding_controller.py", line 38, in _forward_logs     for message in service.read_logs(event.bucket, event.key):   File "/var/task/azul/indexer/log_forwarding_service.py", line 38, in read_logs     for message in self._parse_log_lines(body):   File "/var/task/azul/indexer/log_forwarding_service.py", line 130, in _parse_log_lines     for message in super()._parse_log_lines(file_body):   File "/var/task/azul/indexer/log_forwarding_service.py", line 54, in _parse_log_lines     for row in csv.reader(file_body, delimiter=' ', quotechar='"'):   File "/var/task/azul/indexer/log_forwarding_service.py", line 127, in _read_log     yield line.decode('ascii') |
| 2024-12-14 22:27:02.485 | END RequestId: 7ee248cf-0988-48e1-971e-999432d214a6 |
| 2024-12-14 22:27:02.485 | REPORT RequestId: 7ee248cf-0988-48e1-971e-999432d214a6 Duration: 24.94 ms Billed Duration: 25 ms Memory Size: 128 MB Max Memory Used: 122 MB |
| 2024-12-14 22:29:07.634 | START RequestId: 7ee248cf-0988-48e1-971e-999432d214a6 Version: $LATEST |
| 2024-12-14 22:29:07.636 | [INFO] 2024-12-14T22:29:07.636Z 7ee248cf-0988-48e1-971e-999432d214a6 azul.boto3 s3.GetObject: Making GET request to https://s3.amazonaws.com/edu-ucsc-gi-platform-anvil-prod-logs.us-east-1/s3/access/anvilprod/browser/2024-12-14-22-25-54-E363A95D4EFEAC8A |
| 2024-12-14 22:29:07.636 | [INFO] 2024-12-14T22:29:07.636Z 7ee248cf-0988-48e1-971e-999432d214a6 azul.boto3 s3.GetObject: … without request body |
| 2024-12-14 22:29:07.665 | [INFO] 2024-12-14T22:29:07.665Z 7ee248cf-0988-48e1-971e-999432d214a6 azul.boto3 s3.GetObject: Got 200 response |
| 2024-12-14 22:29:07.665 | [INFO] 2024-12-14T22:29:07.665Z 7ee248cf-0988-48e1-971e-999432d214a6 azul.boto3 s3.GetObject: … with nonprintable body (<class 'botocore.response.StreamingBody'>) |
| 2024-12-14 22:29:07.666 | LAMBDA_WARNING: Unhandled exception. The most likely cause is an issue in the function code. However, in rare cases, a Lambda runtime update can cause unexpected function behavior. For functions using managed runtimes, runtime updates can be triggered by a function change, or can be applied automatically. To determine if the runtime has been updated, check the runtime version in the INIT_START log entry. If this error correlates with a change in the runtime version, you may be able to mitigate this error by temporarily rolling back to the previous runtime version. For more information, see https://docs.aws.amazon.com/lambda/latest/dg/runtimes-update.html |
| 2024-12-14 22:29:07.666 | [ERROR] UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 315: ordinal not in range(128) Traceback (most recent call last):   File "/var/task/azul/chalice.py", line 181, in patched_event_source_handler     return old_handler(self_, event, context)   File "/var/task/chalice/app.py", line 1756, in __call__     return self.handler(event_obj)   File "/var/task/app.py", line 255, in forward_s3_logs     app.log_controller.forward_s3_access_logs(event)   File "/var/task/azul/indexer/log_forwarding_controller.py", line 47, in forward_s3_access_logs     self._forward_logs(event, self.s3)   File "/var/task/azul/indexer/log_forwarding_controller.py", line 38, in _forward_logs     for message in service.read_logs(event.bucket, event.key):   File "/var/task/azul/indexer/log_forwarding_service.py", line 38, in read_logs     for message in self._parse_log_lines(body):   File "/var/task/azul/indexer/log_forwarding_service.py", line 130, in _parse_log_lines     for message in super()._parse_log_lines(file_body):   File "/var/task/azul/indexer/log_forwarding_service.py", line 54, in _parse_log_lines     for row in csv.reader(file_body, delimiter=' ', quotechar='"'):   File "/var/task/azul/indexer/log_forwarding_service.py", line 127, in _read_log     yield line.decode('ascii') |
| 2024-12-14 22:29:07.685 | END RequestId: 7ee248cf-0988-48e1-971e-999432d214a6 |
| 2024-12-14 22:29:07.685 | REPORT RequestId: 7ee248cf-0988-48e1-971e-999432d214a6 Duration: 50.73 ms Billed Duration: 51 ms Memory Size: 128 MB Max Memory Used: 122 MB |
---

--- a/src/azul/indexer/log_forwarding_service.py
+++ b/src/azul/indexer/log_forwarding_service.py
@@ -124,7 +124,15 @@ class S3AccessLogForwardingService(LogForwardingService):
 
     def _read_log(self, response: StreamingBody) -> Iterable[str]:
         for line in response.iter_lines():
-            yield line.decode('ascii')
+            # AWS does not document what encoding is used for these log objects.
+            # Using Latin-1 ensures the log forwarder won't fail due to decoding
+            # errors, since every octet is a valid Latin-1 character. The
+            # strings will be re-encoded using UTF-8, so the binary content of
+            # the CloudWatch log messages will differ from the S3 log if the
+            # latter contains characters that can't be encoded using UTF-8. This
+            # decision was made to minimize time spent diagnosing these decoding
+            # errors.
+            yield line.decode('latin1')
 
     def _parse_log_lines(self, file_body: Iterable[str]) -> Iterator[MutableJSON]:
         for message in super()._parse_log_lines(file_body):

